### ISSUE TITLE
DNC-SFA called via Saihu: Cannot verify provided result/s

### ISSUE THREAD

**Issue by sbondorf** [adfeel220/Saihu-TSN-Analysis-Tool-Integration#10]
The NCorg DNC offers small test networks, for functional tests, see files with ``_ServerGraph.java`` suffix [here](https://github.com/NetCal/DNC_func_tests/tree/master/java/org/networkcalculus/dnc/func_tests).

I picked ``TA_2S_2SC_2F_1AC_1P_ServerGraph.java`` for a small test run. The network consists of a tandem (TA) topology with two servers (2S) having two different service curves (2SC). The network is crossed by two flows (2F) with equal arrival curves (1AC) and equal paths (1P). They cross both servers.

I modeled this in JSON for Saihu, see this file: [TA_2S_2SC_2F_1AC_1P.json](https://github.com/adfeel220/Saihu-TSN-Analysis-Tool-Integration/files/13948453/TA_2S_2SC_2F_1AC_1P.json)

To call the DNC's (arbitrary multiplexing) analyses directly from within the DNC tool, see the code in the next comment (cannot paste Java files here ...). It takes the configurations that Saihu applies to the DNC, copied over from Saihu's NetworkAnalysis class.

Delay bounds of TFA, PMOO and TMA match the expected ones -- i.e., previously manually computed results, results when using DNC directly, and results when calling the DNC via Saihu are all the same. SFA does, however, not. Calling DNC directly and [manually computed](https://github.com/adfeel220/Saihu-TSN-Analysis-Tool-Integration/files/13948583/TA_2S_2SC_2F_1AC_1P_by-hand.pdf) results match (43 1/7); Saihu provides a considerably different result (40 6.5/14).

Am I missing some potentially undocumented configuration that Saihu applies to the DNC's SFA?

**Comment by sbondorf** [comment]
```
import org.networkcalculus.dnc.AnalysisConfig;
import org.networkcalculus.dnc.network.server_graph.Flow;
import org.networkcalculus.dnc.network.server_graph.ServerGraph;
import org.networkcalculus.dnc.tandem.analyses.PmooAnalysis;
import org.networkcalculus.dnc.tandem.analyses.SeparateFlowAnalysis;
import org.networkcalculus.dnc.tandem.analyses.TandemMatchingAnalysis;
import org.networkcalculus.dnc.tandem.analyses.TotalFlowAnalysis;
import org.networkcalculus.num.Num;

public class TA_demo {

    public TA_demo() {
    }

    public static void main(String[] args) {
        TA_demo demo = new TA_demo();

        try {
            demo.run();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void run() throws Exception {
        ServerGraph sg = new TA_2S_2SC_2F_1AC_1P_ServerGraph().getServerGraph();

        // copied from Saihu
        AnalysisConfig config = new AnalysisConfig();
        config.enforceMultiplexing(AnalysisConfig.MultiplexingEnforcement.GLOBAL_ARBITRARY);

        Num totalDelay;
        Num maxBacklog;
        long startTime;
        long exec_time;

        for(Flow foi : sg.getFlows()) {

            System.out.println( "foi " + foi.getAlias());

            System.out.println( "TFA");
            // copied from Saihu
            config.enforceMaxSC(AnalysisConfig.MaxScEnforcement.GLOBALLY_OFF);
            config.enforceMaxScOutputRate(AnalysisConfig.MaxScEnforcement.GLOBALLY_OFF);
            config.setArrivalBoundMethod(AnalysisConfig.ArrivalBoundMethod.AGGR_PBOO_CONCATENATION);
            TotalFlowAnalysis tfa = new TotalFlowAnalysis(sg, config);

            startTime = System.currentTimeMillis();
            tfa.performAnalysis(foi);
            exec_time = System.currentTimeMillis() - startTime;

            totalDelay = tfa.getDelayBound();
            maxBacklog = tfa.getBacklogBound();

            System.out.println( "total_delay " + Double.parseDouble(totalDelay.toString()) );
            System.out.println( "max_backlog " + Double.parseDouble(maxBacklog.toString()) );
            System.out.println( "exec_time " + exec_time);
            System.out.println();



            System.out.println( "SFA");
            // copied from Saihu
            config.enforceMaxSC(AnalysisConfig.MaxScEnforcement.GLOBALLY_OFF);
            config.enforceMaxScOutputRate(AnalysisConfig.MaxScEnforcement.GLOBALLY_OFF);
            config.setArrivalBoundMethod(AnalysisConfig.ArrivalBoundMethod.SEGR_PBOO);
            SeparateFlowAnalysis sfa = new SeparateFlowAnalysis(sg, config);

            startTime = System.currentTimeMillis();
            sfa.performAnalysis(foi);
            exec_time = System.currentTimeMillis() - startTime;

            totalDelay = sfa.getDelayBound();
            maxBacklog = sfa.getBacklogBound();

            System.out.println( "total_delay " + Double.parseDouble(totalDelay.toString()) );
            System.out.println( "max_backlog " + Double.parseDouble(maxBacklog.toString()) );
            System.out.println( "exec_time " + exec_time);
            System.out.println();



            System.out.println( "PMOO");
            // copied from Saihu
            config.enforceMultiplexing(AnalysisConfig.MultiplexingEnforcement.GLOBAL_ARBITRARY);
            config.enforceMaxSC(AnalysisConfig.MaxScEnforcement.GLOBALLY_OFF);
            config.enforceMaxScOutputRate(AnalysisConfig.MaxScEnforcement.GLOBALLY_OFF);
            config.setArrivalBoundMethod(AnalysisConfig.ArrivalBoundMethod.AGGR_PMOO);
            PmooAnalysis pmoo = new PmooAnalysis(sg, config);

            startTime = System.currentTimeMillis();
            pmoo.performAnalysis(foi);
            exec_time = System.currentTimeMillis() - startTime;

            totalDelay = pmoo.getDelayBound();
            maxBacklog = pmoo.getBacklogBound();

            System.out.println( "total_delay " + Double.parseDouble(totalDelay.toString()) );
            System.out.println( "max_backlog " + Double.parseDouble(maxBacklog.toString()) );
            System.out.println( "exec_time " + exec_time);
            System.out.println();


            System.out.println( "TMA");
            // copied from Saihu
            config.enforceMultiplexing(AnalysisConfig.MultiplexingEnforcement.GLOBAL_ARBITRARY);
            config.enforceMaxSC(AnalysisConfig.MaxScEnforcement.GLOBALLY_OFF);
            config.enforceMaxScOutputRate(AnalysisConfig.MaxScEnforcement.GLOBALLY_OFF);
            config.setArrivalBoundMethod(AnalysisConfig.ArrivalBoundMethod.AGGR_TM);
            TandemMatchingAnalysis tma = new TandemMatchingAnalysis(sg, config);

            startTime = System.currentTimeMillis();
            tma.performAnalysis(foi);
            exec_time = System.currentTimeMillis() - startTime;

            totalDelay = tma.getDelayBound();
            maxBacklog = tma.getBacklogBound();

            System.out.println( "total_delay " + Double.parseDouble(totalDelay.toString()) );
            System.out.println( "max_backlog " + Double.parseDouble(maxBacklog.toString()) );
            System.out.println( "exec_time " + exec_time);

            System.out.println();
            System.out.println();
        }
    }
}
```

**Comment by adfeel220** [comment]
Hi, sorry for the late reply. I was overwhelmed by another project in the past month.

# TL;DR
I suppose PR #11 should fix the issue.

# Reason why it failed
The problem is that in DNC, I thought the local multiplexing scheme was overwritten by the global multiplexing setting. It turns out that if one adds a server with a pre-set multiplexing scheme, the following global enforcement does not apply to the servers. 

For example, if I set the multiplexing locally to FIFO on a server,
```java
network.addServer("server1", serviceCurve, maxServiceCurve, AnalysisConfig.Multiplexing.FIFO);
```
Then apply the network to the following analysis.
```java
config.enforceMultiplexing(AnalysisConfig.MultiplexingEnforcement.GLOBAL_ARBITRARY);
SeparateFlowAnalysis sfa = new SeparateFlowAnalysis(sg, config);
```

The service curve is still computed as if with FIFO, that was the reason why Saihu returned a slightly smaller number. I am not sure if it's an expected behavior in DNC. Nevertheless, I changed the default server multiplexing to `Arbitrary` after checking that DNC sets arbitrary multiplexing as default.

# Remark
Thank you for your feedback and feel free to point out any problems you may encounter! Most of the time I may miss some special testing scenarios so any feedback is appreciated.
